platform/storage/posix/src/mender-storage.c

```diff
@@ -151,24 +151,101 @@ mender_storage_delete_authentication_keys(void) {
+     * the format of the latest and greatest version. By updating the version
+     * number here and validating the data below, we make sure to enforce this.
+     */
+    cJSON *item;
+    if (NULL == (item = cJSON_CreateNumber((double)MENDER_STORAGE_DEPLOYMENT_DATA_VERSION))) {
+        mender_log_error("Unable to allocate memory");
+        return MENDER_FAIL;
+    }
+    if (NULL == cJSON_GetObjectItem(deployment_data, "version")) {
+        if (!cJSON_AddItemToObject(deployment_data, "version", item)) {
+            mender_log_error("Unable to allocate memory");
+            cJSON_Delete(item);
+            return MENDER_FAIL;
+        }
+    } else {
+        /* One would expect cJSON_AddItemToObject above to replace the value on
+         * already existing keys. However, it creates another entry with a
+         * duplicate key, which is invalid JSON. So we have to be careful. */
+        if (!cJSON_ReplaceItemInObject(deployment_data, "version", item)) {
+            mender_log_error("Unable to allocate memory");
+            cJSON_Delete(item);
+            return MENDER_FAIL;
+        }
+    }
+
+    /* Validate deployment data */
+    if (!mender_utils_validate_deployment_data(deployment_data, MENDER_STORAGE_DEPLOYMENT_DATA_VERSION)) {
+        mender_log_error("Invalid deployment data");
+        return MENDER_FAIL;
+    }
+
+    /* Compose JSON string */
+    char *json_str;
+    if (NULL == (json_str = cJSON_PrintUnformatted(deployment_data))) {
+        mender_log_error("Unable to compose deployment data");
+        return MENDER_FAIL;
+    }
+
+    /* Write to store */
+    size_t json_str_len = strlen(json_str);
+    if (MENDER_OK != mender_storage_write_file(MENDER_STORAGE_NVS_DEPLOYMENT_DATA, json_str, json_str_len)) {
+        mender_log_error("Unable store deployment data");
+        free(json_str);
         return MENDER_FAIL;
     }
+
+    free(json_str);
     return MENDER_OK;
 }
 
 mender_err_t
-mender_storage_get_deployment_data(char **deployment_data) {
+mender_storage_get_deployment_data(cJSON **deployment_data) {
     assert(NULL != deployment_data);
 
-    size_t deployment_data_length;
-    if (MENDER_OK != mender_storage_read_file(MENDER_STORAGE_NVS_DEPLOYMENT_DATA, (void **)deployment_data, &deployment_data_length)) {
-        return MENDER_NOT_FOUND;
+    char  *json_str;
+    size_t json_str_len;
+
+    /* Get deployment data from store. We differentiate between error and not
+     * found. Not found may be OK. */
+    mender_err_t ret;
+    switch (ret = mender_storage_read_file(MENDER_STORAGE_NVS_DEPLOYMENT_DATA, (void **)&json_str, &json_str_len)) {
+        case MENDER_OK:
+            break;
+        case MENDER_NOT_FOUND:
+            mender_log_debug("Deployment data not available");
+            return ret;
+        case MENDER_DONE:
+        case MENDER_NOT_IMPLEMENTED:
+        case MENDER_FAIL:
+            mender_log_error("Unable to read deployment data");
+            return ret;
```

Comment:
Shouldn't we always return `MENDER_FAIL` in this last case? The caller of this function should not bother about the various failure cases, right?

Reply:
This is no longer relevant, as mender-storage.c is effectively untouched now. But yes, I agree 👍